sympy/simplify/simplify.py

```diff
@@ -514,8 +515,18 @@ def simplify(expr, ratio=1.7, measure=count_ops, rational=False, inverse=False,
     expression. You can avoid this behavior by passing ``doit=False`` as
     an argument.
     """
+
+    def shorter(*choices):
+        """
+        Return the choice that has the fewest ops. In case of a tie,
+        the expression listed first is selected.
+        """
+        if not has_variety(choices):
+            return choices[0]
+        return min(choices, key=measure)
+
     def done(e):
-        return e.doit() if doit else e
+        return shorter(e.doit(), e) if doit else e
```

Comment:
As you noticed [here](https://github.com/sympy/sympy/pull/17259/files#r307408318), allowing selection between the doit/undoit form does not respect the doit flag. The docstring says that doit will be called if the flag is not False. If you make it so that the selection is only done if doit is None then a routine can request the shortest and accept the undoit form for compactness sake:
```python
return e.doit() if doit else (shorter(e.doit(), e) if doit is None else e)
```
If so, the docstring should be updated.

Reply:
The reason for this is https://github.com/sympy/sympy/issues/17165#issuecomment-514748083

But you are right that the docstring should be updated. (If it wasn't for the expression growth, I would not have touched it, as I think it makes sense to doit.)

Reply:
The concensus was formerly that doit=True was the most user-friendly setting; I would recommend calling with doit=None from Piecewise.

Reply:
OK! Not sure if I follow though. In the internal `_eval_simplify` in Piecewise, I call simplify with doit=None instead of passing the existing doit-flag? (On the expression, I doubt something like that is likely to happen for the conditions.) I guess that will solve the problem, although one cannot force doit on the Piecewise expressions then. But I guess that is the price to pay (and better than the expression growth...). Will try it out.

Reply:
It doesn't work. In the Piecewise specific code, I call simplify on each part and get this:

```
s = (x*Sum(x**n, (n, -1, oo)))
s.doit()
```
![image](https://user-images.githubusercontent.com/8114497/61978054-72876b00-aff0-11e9-9407-7c8f534dbd90.png)
```
s.doit().simplify()
```
![image](https://user-images.githubusercontent.com/8114497/61978069-8206b400-aff0-11e9-9278-9b75b2f4574e.png)
```
s.doit().simplify().simplify()
```
![image](https://user-images.githubusercontent.com/8114497/61978090-93e85700-aff0-11e9-9953-0960882e8dba.png)

From here on it starts looping and keeps constant.

Or did you mean to check if the expression in simplify is a Piecewise and in that case not doit?

Reply:
Oh, and this includes an additional piecewise_fold after the piecewise specific part. So I am quite sure that the final doit is done either way.

Reply:
What about applying a piecewise_fold after the final doit if it contains a Piecewise?

Reply:
Maybe...

Reply:
That seems to do the trick. I'm actually wondering if one should have a special path for Piecewise expressions? Not sure if any of the later stages does anything with a Piecewise anyway, so maybe better to just run simplify on the expressions individually (after a piecewise_fold), and then try to simplify it from a Piecewise-perspective (getting rid of pieces).